File: src/shared/wire.ts
export type PayloadType = "image" | "text";

export interface PayloadFrame {
  v: 1;
  type: PayloadType;
  mime: string;
  origin: string;
  ts: number;
  nonce: string;
  payload: string;
}

export interface PayloadMetadata {
  type: PayloadType;
  mime: string;
  origin: string;
  ts: number;
}

export interface RelayHealth {
  status: "ok" | "degraded";
  relayName: string;
  clipboard?: {
    enabled: boolean;
    status: "starting" | "healthy" | "degraded" | "disabled";
    watcher?: string;
    error?: string;
  };
}

export type TransferDirection = "laptop_to_phone" | "phone_to_laptop";

export interface TransferFileOffer {
  fileId: string;
  filename: string;
  mime: string;
  size: number;
  lastModifiedMs: number;
  sha256: string;
}

export interface TransferOffer {
  transferId: string;
  batchId: string;
  origin: string;
  direction: TransferDirection;
  createdAtMs: number;
  files: TransferFileOffer[];
}

export type TransferControlMessage =
  | { v: 1; kind: "transfer_offer"; offer: TransferOffer }
  | {
      v: 1;
      kind: "transfer_accept";
      transferId: string;
      fileId: string;
      confirmedOffset: number;
    }
  | {
      v: 1;
      kind: "transfer_progress";
      transferId: string;
      fileId: string;
      confirmedOffset: number;
    }
  | {
      v: 1;
      kind: "transfer_pause" | "transfer_resume" | "transfer_cancel";
      transferId: string;
      fileId?: string;
    }
  | {
      v: 1;
      kind: "transfer_file_complete";
      transferId: string;
      fileId: string;
      sha256: string;
    }
  | {
      v: 1;
      kind: "transfer_file_failed";
      transferId: string;
      fileId: string;
      code: string;
    };

export type RelayMessage =
  | { v: 1; kind: "hello"; challenge: string; maxPayloadBytes: number }
  | { v: 1; kind: "auth"; deviceId: string; proof: string }
  | { v: 1; kind: "auth_ok"; health?: RelayHealth }
  | { v: 1; kind: "health"; health: RelayHealth }
  | { v: 1; kind: "publish"; frame: PayloadFrame }
  | { v: 1; kind: "payload"; frame: PayloadFrame }
  | { v: 1; kind: "ack"; ts: number }
  | TransferControlMessage
  | { v: 1; kind: "error"; code: string; message: string };

export function isPayloadFrame(value: unknown): value is PayloadFrame {
  if (!value || typeof value !== "object") return false;
  const frame = value as Record<string, unknown>;
  return (
    frame.v === 1 &&
    (frame.type === "image" || frame.type === "text") &&
    typeof frame.mime === "string" &&
    typeof frame.origin === "string" &&
    typeof frame.ts === "number" &&
    Number.isFinite(frame.ts) &&
    typeof frame.nonce === "string" &&
    typeof frame.payload === "string"
  );
}

export function encodedPayloadBytes(frame: PayloadFrame): number {
  return Buffer.from(frame.payload, "base64").byteLength;
}

const idPattern = /^[A-Za-z0-9_-]{16,128}$/;
const sha256Pattern = /^[a-f0-9]{64}$/;

export function isTransferOffer(value: unknown): value is TransferOffer {
  if (!value || typeof value !== "object") return false;
  const offer = value as Record<string, unknown>;
  return (
    isId(offer.transferId) &&
    isId(offer.batchId) &&
    typeof offer.origin === "string" &&
    offer.origin.length > 0 &&
    (offer.direction === "laptop_to_phone" ||
      offer.direction === "phone_to_laptop") &&
    isNonNegativeSafeInteger(offer.createdAtMs) &&
    Array.isArray(offer.files) &&
    offer.files.length > 0 &&
    offer.files.every(isTransferFileOffer) &&
    new Set(
      offer.files.map((file) => (file as TransferFileOffer).fileId),
    ).size === offer.files.length
  );
}

export function isTransferControlMessage(
  value: unknown,
): value is TransferControlMessage {
  if (!value || typeof value !== "object") return false;
  const message = value as Record<string, unknown>;
  if (message.v !== 1 || typeof message.kind !== "string") return false;

  switch (message.kind) {
    case "transfer_offer":
      return isTransferOffer(message.offer);
    case "transfer_accept":
    case "transfer_progress":
      return (
        isId(message.transferId) &&
        isId(message.fileId) &&
        isNonNegativeSafeInteger(message.confirmedOffset)
      );
    case "transfer_pause":
    case "transfer_resume":
    case "transfer_cancel":
      return (
        isId(message.transferId) &&
        (message.fileId === undefined || isId(message.fileId))
      );
    case "transfer_file_complete":
      return (
        isId(message.transferId) &&
        isId(message.fileId) &&
        typeof message.sha256 === "string" &&
        sha256Pattern.test(message.sha256)
      );
    case "transfer_file_failed":
      return (
        isId(message.transferId) &&
        isId(message.fileId) &&
        typeof message.code === "string" &&
        message.code.length > 0
      );
    default:
      return false;
  }
}

function isTransferFileOffer(value: unknown): value is TransferFileOffer {
  if (!value || typeof value !== "object") return false;
  const file = value as Record<string, unknown>;
  return (
    isId(file.fileId) &&
    typeof file.filename === "string" &&
    file.filename.length > 0 &&
    file.filename.length <= 255 &&
    !file.filename.includes("/") &&
    !file.filename.includes("\\") &&
    file.filename !== "." &&
    file.filename !== ".." &&
    typeof file.mime === "string" &&
    file.mime.length > 0 &&
    isNonNegativeSafeInteger(file.size) &&
    isNonNegativeSafeInteger(file.lastModifiedMs) &&
    typeof file.sha256 === "string" &&
    sha256Pattern.test(file.sha256)
  );
}

function isId(value: unknown): value is string {
  return typeof value === "string" && idPattern.test(value);
}

function isNonNegativeSafeInteger(value: unknown): value is number {
  return (
    typeof value === "number" &&
    Number.isSafeInteger(value) &&
    value >= 0
  );
}

File: app/lib/src/shared/wire.dart
import 'dart:convert';

enum PayloadType {
  image('image'),
  text('text');

  const PayloadType(this.wireName);

  final String wireName;

  static PayloadType parse(String value) {
    return PayloadType.values.firstWhere(
      (type) => type.wireName == value,
      orElse: () => throw FormatException('Unsupported payload type: $value'),
    );
  }
}

class PayloadMetadata {
  const PayloadMetadata({
    required this.type,
    required this.mime,
    required this.origin,
    required this.ts,
  });

  final PayloadType type;
  final String mime;
  final String origin;
  final int ts;
}

class PayloadFrame {
  const PayloadFrame({
    required this.type,
    required this.mime,
    required this.origin,
    required this.ts,
    required this.nonce,
    required this.payload,
    this.v = 1,
  });

  final int v;
  final PayloadType type;
  final String mime;
  final String origin;
  final int ts;
  final String nonce;
  final String payload;

  @override
  bool operator ==(Object other) {
    return other is PayloadFrame &&
        other.v == v &&
        other.type == type &&
        other.mime == mime &&
        other.origin == origin &&
        other.ts == ts &&
        other.nonce == nonce &&
        other.payload == payload;
  }

  @override
  int get hashCode => Object.hash(v, type, mime, origin, ts, nonce, payload);

  Map<String, Object?> toJson() => {
    'v': v,
    'type': type.wireName,
    'mime': mime,
    'origin': origin,
    'ts': ts,
    'nonce': nonce,
    'payload': payload,
  };

  Map<String, Object?> associatedDataJson() => {
    'v': v,
    'type': type.wireName,
    'mime': mime,
    'origin': origin,
    'ts': ts,
    'nonce': nonce,
  };

  String associatedData() => jsonEncode(associatedDataJson());

  PayloadFrame copyWith({
    int? v,
    PayloadType? type,
    String? mime,
    String? origin,
    int? ts,
    String? nonce,
    String? payload,
  }) {
    return PayloadFrame(
      v: v ?? this.v,
      type: type ?? this.type,
      mime: mime ?? this.mime,
      origin: origin ?? this.origin,
      ts: ts ?? this.ts,
      nonce: nonce ?? this.nonce,
      payload: payload ?? this.payload,
    );
  }

  static PayloadFrame fromJson(Object? value) {
    if (value is! Map<String, Object?>) {
      throw const FormatException('Payload frame must be an object.');
    }
    return PayloadFrame(
      v: _intField(value, 'v'),
      type: PayloadType.parse(_stringField(value, 'type')),
      mime: _stringField(value, 'mime'),
      origin: _stringField(value, 'origin'),
      ts: _intField(value, 'ts'),
      nonce: _stringField(value, 'nonce'),
      payload: _stringField(value, 'payload'),
    );
  }
}

enum TransferDirection {
  laptopToPhone('laptop_to_phone'),
  phoneToLaptop('phone_to_laptop');

  const TransferDirection(this.wireName);

  final String wireName;

  static TransferDirection parse(String value) {
    return TransferDirection.values.firstWhere(
      (direction) => direction.wireName == value,
      orElse: () =>
          throw FormatException('Unsupported transfer direction: $value'),
    );
  }
}

class TransferFileOffer {
  const TransferFileOffer({
    required this.fileId,
    required this.filename,
    required this.mime,
    required this.size,
    required this.lastModifiedMs,
    required this.sha256,
  });

  final String fileId;
  final String filename;
  final String mime;
  final int size;
  final int lastModifiedMs;
  final String sha256;

  Map<String, Object?> toJson() => {
    'fileId': fileId,
    'filename': filename,
    'mime': mime,
    'size': size,
    'lastModifiedMs': lastModifiedMs,
    'sha256': sha256,
  };

  static TransferFileOffer fromJson(Object? value) {
    if (value is! Map) {
      throw const FormatException('Transfer file offer must be an object.');
    }
    final json = value.cast<Object?, Object?>();
    final fileId = _transferIdField(json, 'fileId');
    final filename = _stringField(json, 'filename');
    if (filename.length > 255 ||
        filename == '.' ||
        filename == '..' ||
        filename.contains('/') ||
        filename.contains(r'\')) {
      throw const FormatException('filename must be a safe basename.');
    }
    final size = _nonNegativeIntField(json, 'size');
    final lastModifiedMs = _nonNegativeIntField(json, 'lastModifiedMs');
    final sha256 = _stringField(json, 'sha256');
    if (!_sha256Pattern.hasMatch(sha256)) {
      throw const FormatException('sha256 must be lowercase hexadecimal.');
    }
    return TransferFileOffer(
      fileId: fileId,
      filename: filename,
      mime: _stringField(json, 'mime'),
      size: size,
      lastModifiedMs: lastModifiedMs,
      sha256: sha256,
    );
  }
}

class TransferOffer {
  const TransferOffer({
    required this.transferId,
    required this.batchId,
    required this.origin,
    required this.direction,
    required this.createdAtMs,
    required this.files,
  });

  final String transferId;
  final String batchId;
  final String origin;
  final TransferDirection direction;
  final int createdAtMs;
  final List<TransferFileOffer> files;

  Map<String, Object?> toJson() => {
    'transferId': transferId,
    'batchId': batchId,
    'origin': origin,
    'direction': direction.wireName,
    'createdAtMs': createdAtMs,
    'files': files.map((file) => file.toJson()).toList(),
  };

  static TransferOffer fromJson(Object? value) {
    if (value is! Map) {
      throw const FormatException('Transfer offer must be an object.');
    }
    final json = value.cast<Object?, Object?>();
    final rawFiles = json['files'];
    if (rawFiles is! List || rawFiles.isEmpty) {
      throw const FormatException('files must be a non-empty list.');
    }
    final files = rawFiles.map(TransferFileOffer.fromJson).toList();
    if (files.map((file) => file.fileId).toSet().length != files.length) {
      throw const FormatException('fileId must be unique within a transfer.');
    }
    return TransferOffer(
      transferId: _transferIdField(json, 'transferId'),
      batchId: _transferIdField(json, 'batchId'),
      origin: _stringField(json, 'origin'),
      direction: TransferDirection.parse(_stringField(json, 'direction')),
      createdAtMs: _nonNegativeIntField(json, 'createdAtMs'),
      files: files,
    );
  }
}

final _transferIdPattern = RegExp(r'^[A-Za-z0-9_-]{16,128}$');
final _sha256Pattern = RegExp(r'^[a-f0-9]{64}$');

String _transferIdField(Map<Object?, Object?> json, String field) {
  final value = _stringField(json, field);
  if (!_transferIdPattern.hasMatch(value)) {
    throw FormatException('$field must be a valid transfer identifier.');
  }
  return value;
}

int _nonNegativeIntField(Map<Object?, Object?> json, String field) {
  final value = json[field];
  if (value is! int || value < 0) {
    throw FormatException('$field must be a non-negative integer.');
  }
  return value;
}

int _intField(Map<String, Object?> json, String field) {
  final value = json[field];
  if (value is! int) throw FormatException('$field must be an integer.');
  return value;
}

String _stringField(Map<Object?, Object?> json, String field) {
  final value = json[field];
  if (value is! String || value.isEmpty) {
    throw FormatException('$field must be a string.');
  }
  return value;
}
